internal/runners/builds/download.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/ActiveState/cli/pkg/platform/runtime/setup/events"
+	"github.com/ActiveState/cli/pkg/platform/runtime/setup/events/progress"
+	"github.com/ActiveState/cli/pkg/platform/runtime/store"
+	"github.com/ActiveState/cli/pkg/platform/runtime/target"
+	"github.com/ActiveState/cli/pkg/project"
+	"github.com/go-openapi/strfmt"
+)
+
+type DownloadParams struct {
+	BuildID   string
+	OutputDir string
+}
+
+type Download struct {
+	out       output.Outputer
+	project   *project.Project
+	analytics analytics.Dispatcher
+	svcModel  *model.SvcModel
+	auth      *auth.Auth
+	config    *config.Instance
+}
+
+func NewDownload(prime primeable) *Download {
+	return &Download{
+		out:       prime.Output(),
+		project:   prime.Project(),
+		analytics: prime.Analytics(),
+		svcModel:  prime.SvcModel(),
+		auth:      prime.Auth(),
+		config:    prime.Config(),
+	}
+}
+
+func (d *Download) Run(params *DownloadParams) (rerr error) {
+	defer rationalizeError(&rerr)
+
+	if d.project == nil {
+		return rationalize.ErrNoProject
+	}
+
+	pg := runbits.NewRuntimeProgressIndicator(d.out)
+	defer rtutils.Closer(pg.Close, &rerr)
+
+	if err := pg.Handle(events.SolveStart{}); err != nil {
+		return errs.Wrap(err, "Failed to handle SolveStart event")
+	}
+
+	_, err := runtime.NewFromProject(d.project, target.TriggerBuilds, d.analytics, d.svcModel, d.out, d.auth, d.config)
+	if err != nil {
+		return locale.WrapInputError(err, "err_refresh_runtime_new", "Could not update runtime for this project.")
+	}
+
+	runtimeStore := store.New(target.NewProjectTarget(d.project, nil, target.TriggerBuilds).Dir())
+	bp, err := runtimeStore.BuildPlan()
+	if err != nil {
+		return errs.Wrap(err, "Could not get build plan")
+	}
+
+	terminalArtfMap, err := buildplan.NewMapFromBuildPlan(bp, false, false, nil)
+	if err != nil {
+		return errs.Wrap(err, "Could not get build plan map")
+	}
+
+	if err := pg.Handle(events.SolveSuccess{}); err != nil {
+		return errs.Wrap(err, "Failed to handle SolveSuccess event")
+	}
+
+	var artifact *artifact.Artifact
+	for _, artifacts := range terminalArtfMap {
+		for _, a := range artifacts {
+			if strings.Contains(strings.ToLower(string(a.ArtifactID)), strings.ToLower(params.BuildID)) {
+				artifact = &a
+				break
+			}
+		}
+	}
+
+	if artifact == nil {
+		return locale.WrapInputError(err, "err_build_id_not_found", "Could not find build ID {{.V0}}", params.BuildID)
+	}
+
+	if err := d.downloadArtifact(pg, artifact, params.OutputDir); err != nil {
+		return errs.Wrap(err, "Could not download artifact %s", artifact.ArtifactID.String())
+	}
+
+	d.out.Notice(locale.Tl("msg_download_success", "Downloaded {{.V0}} to {{.V1}}", artifact.ArtifactID.String(), params.OutputDir))
+	return nil
+}
+
+func (d *Download) downloadArtifact(pg events.Handler, artifact *artifact.Artifact, targetDir string) (rerr error) {
+	defer func() {
+		evs := []events.Eventer{
+			events.ArtifactDownloadSuccess{artifact.ArtifactID},
+			events.Success{},
```

Comment:
For some reason after a successful download the artifact bar is completed but the overall download bar is not. I thought this event would handle it but it doesn't. Any insights here would be welcome or I can continue to debug what's going on.